### ISSUE TITLE
RFC: Appending text/files automatically, Auto-Insert

### ISSUE THREAD

**Issue by rqelibari** [nodemailer/wildduck#191]
Hi everyone,

this is a request for comment on a new feature we want to implement.

## RFC: Auto-Insert
We, at [Astzweig](/astzweig), want to be able to automatically append a signature (with or without files) to some of our outgoing emails in our [wildduck](/nodemailer/wildduck) + [ZonteMTA](/zone-eu/zone-mta) mailserver setup.

*Normally this would just concern ZonteMTA (the outbound SMTP server), but as in most setups it is wildduck that stores user configurations and addresses (ZoneMTA just uses the [zonemta-wildduck](/nodemailer/zonemta-wildduck) plugin to access those information), we think wildduck should also be the one storing Auto-Insert related configuration.*

### Assumptions about the relationship of wildduck and ZoneMTA
If I understand correctly there are two ways to send email messages in a wildduck + ZoneMTA setup:

1. Mail clients connect directly to ZoneMTA and ZoneMTA uses zonemta-wildduck plugin to check the configuration data.

2. wildduck gets an API call with message data, that it writes to the ZoneMTA queue in the MongoDB database. It does some validation before though.

In both cases ZoneMTA should be responsible for modifying the message in my opinion. What do you think?

### Our Proposal
Based on our assumptions above, we propose changing the system at two points:

1. We need to extend wildduck to store Auto-Insert settings for each user. We propose extending the user database to hold an array with `auto-insert` information using the `headers`, `text`, `html` and `attachments` key as defined by the submit API of wildduck, using the `query` key as defined by the filter API of wildduck and a new field called `insertAt':

```json
{
  "inserAt": "'start', 'end' or 'beforeLastQuote'",
  "query": "...",
  "headers": "...",
  "text": "...",
  "html": "...",
  "attachments": "..."
}
```

2. We need to extend the zonemta-wildduck plugin to give ZoneMTA access to the configuration data above and to pre- or append the desired content.

What do you guys think?
@andris9  Do you see any mistake in our assumptions or logical flaws in our proposition?

**Comment by andris9** [comment]
I can see the need for such thing but I would rather go with some kind of a plugin system where you could listen for the mail store hooks and then, if needed, process the email before storing it. The immediate upside would be that I'd be able to move current GPG encryption from core into a plugin.

There are 2 kinds of email stores in use. One is the normal one and the other is outbound queue. Either case once message is stored you can not process the content anymore. For normal store you can't modify it as emails are immutable and for outbound queue you can't touch the message body because at that time DKIM body hash has been already calculated and changing the body would break the hash. So any kind of processing must be done before storage. 

It could be done by providing a message store hook for plugins (for a plugin system that does not exist yet) and some metadata (which kind of storage is it, which service is storing the message, etc).The hook would get parsed headers object and the entire body (without headers) of the message as a buffer. Hook would be able to manipulate headers and it would return also a buffer. Whatever the hook returns is then sent to the store.

Downside would be (especially when using multiple plugins) that the same message would have to be parsed several times but for first PoC I guess that would be OK as there actually is no good unserialized email format.

**Comment by rqelibari** [comment]
@andris9 Thank you for your feedback!

I have some follow-up questions:

1. I do not quite understand, why a new plugin system is needed. There is already a plugin systems for Zone-MTA. Can't that just be extended? There is a ['message:store'](/zone-eu/zone-mta/tree/master/plugins) hook for example. It does not allow to modify the body, but we could extend Zone-MTA with another hook called 'message:before_store' for example.

2. What is the 'normal' store? Do you mean the wildduck database (e.g. the sent folder)?

3.  If I understand correctly, the [mail user agent](https://en.wikipedia.org/wiki/Email_client) (e.g. Outlook) connects to Zone-MTA on port 587, which is the [message submission agent](https://en.wikipedia.org/wiki/Message_submission_agent). Zone-MTA uses the zonemta-wildduck plugin, to authenenticate the user and accepts the message, before it puts the message into the outbound queue store. At the same time the zonemta-wildduck hooks into the 'message:store' hook - as mentioned above - and stores a copy of the message in the 'wildduck' database in the users sent folder. Is that correct?

**Comment by rqelibari** [comment]
@andris9 just a ping.

**Comment by andris9** [comment]
Sorry for the delay.

1. ZoneMTA has its plugin system and it works **but** the mail processing plugins only affect emails sent through SMTP. WildDuck itself does not use SMTP at all and rather just inserts all messages directly to the outbound queue bypassing ZoneMTA message processing pipeline. That are: all messages sent through webmail and API, autoreplies, bounces, auto-forwards etc. Patching zone-mta plugin system to work also with WildDuck is quite complicated as the message format does not match.

2. "normal" store is the reqular mail store, accessible via IMAP and SMTP. Outbound queue is also a mail store, only for outgoing SMTP, both store rfc822 emails. And, if processed through API, there is not too much difference between these two, so if already processing messages inserted to outbound queue then allowing to do the same with normal messages would probably make sense.

3. Outlook connects to ZoneMTA, sends a message through SMTP and then **does the same** over IMAP, so messages are actually uploaded twice. Gmail server uploads all messages from SMTP automatically to IMAP store as well, so I tried to do the same in WildDuck but it turned out to be too difficult - messages uploaded to SMTP and IMAP are often different, Gmail manages to figure out the duplicates and only keep a single message copy (uploading via IMAP overrides the message automatically appended from SMTP) but I did not. There were too many duplicates, users started to complain, so I disabled the functionality.
